### ISSUE TITLE
IllegalArgumentException with "pretty printed" fields and raw SQL fetches

### ISSUE THREAD

**Issue by arsinio** [jOOQ/jOOQ#12082]
### Steps to reproduce the problem
Main MCVE is here:
```java
package org.jooq.mcve.test.java;

import org.jooq.mcve.java.tables.records.TestRecord;
import org.junit.Test;
import org.jooq.Field;
import org.jooq.Record;
import org.jooq.Result;
import org.jooq.conf.ParamType;
import org.jooq.impl.DSL;

import static org.jooq.mcve.java.Tables.TEST;
import static org.junit.Assert.assertEquals;

public class JavaTest extends AbstractTest {

   @Test
   public void mcveTest() {
      Field<Integer> field_test = DSL.greatest(
         TEST.ID,
         TEST.VALUE
      );
      String sql = ctx.select(field_test)
                        .from(TEST)
                        .getSQL(ParamType.NAMED_OR_INLINED);
      System.out.println(sql);
      Result<Record> res = ctx.fetch(sql);

      System.out.println(res);

      for( Record currRec : res )
      {
         // this does not work
         System.out.println(currRec.get(field_test));
      }
   }
}
```
Will comment with link to MCVE.

### Expected behavior
`currRec.get(field_test)` should not cause an exception

### Actual behavior
`currRec.get(field_test)` causes an `IllegalArgumentException`
```
java.lang.IllegalArgumentException: Field (greatest(
  "MCVE"."TEST"."ID",
  "MCVE"."TEST"."VALUE"
)) is not contained in Row ("GREATEST(MCVE.TEST.ID, MCVE.TEST.VALUE)")
	at org.jooq.impl.Tools.indexOrFail(Tools.java:1814)
	at org.jooq.impl.AbstractRecord.get(AbstractRecord.java:254)
	at org.jooq.mcve.test.java.JavaTest.mcveTest(JavaTest.java:33)
	...
```
Field names / content seem to match, but I _think_ the `.get()` tries to use a "pretty printed" version of the field with newlines and double quotes.

### Why am I rendering to a SQL string first?
Sometimes we modify the SQL to allow for backward compatibility with removed fields in our database (so we don't need to upload new binaries for older releases). That procedure (not included in the MCVE) shouldn't affect this issue and was removed from the MCVE for clarity.

### Code Generation Configuration
See MCVE

### Versions
- jOOQ: 3.12.0 (earliest tested) - 3.14.11 (latest tested)
- Java: 1.8
- Database (include vendor): H2 included in MCVE, but also occurs in MySQL 5.7.33
- OS: macOS 11.4

**Comment by arsinio** [comment]
MCVE is here:
https://github.com/arsinio/jOOQ-mcve

**Comment by lukaseder** [comment]
Thanks a lot for your detailed report. Will look into this soon.

**Comment by lukaseder** [comment]
I can reproduce the problem once I've added a record to the MCVE database:

```sql
INSERT INTO mcve.test (value) VALUES (1);
```

**Comment by lukaseder** [comment]
Ah yes, I see the point. I don't think you can expect this to work. The field names produced by plain SQL result sets are not reliable, unless you explicitly alias them. Check out what PostgreSQL does for the same kind of query (arguments are not included, only the function name):

![image](https://user-images.githubusercontent.com/734593/123963353-a7f21b80-d9b2-11eb-9b62-6a68f395a669.png)

Or Oracle (arguments are included, but whitespace is stripped, but quotes aren't stripped as in your H2 example):

![image](https://user-images.githubusercontent.com/734593/123963778-1505b100-d9b3-11eb-96a4-ccb96b6a2c33.png)

SQL Server doesn't maintain any column names at all:

![image](https://user-images.githubusercontent.com/734593/123963991-4c745d80-d9b3-11eb-9feb-45ba24abf1f7.png)

Others may produce some sort of indexed column name. This can't be fixed, nor should it be. If you use jOOQ to generate a SQL string, the SQL string no longer "remembers" how it was built, and the result set does not contain any column references you used to build the string.

In fact, numerous other problems will arise if you use this approach, since jOOQ is losing a little bit of information about the original query that was executed (e.g. how many columns there are, if any esoteric feature was used that requires some transformation when fetching data, etc.). You can restore some of that information by using `DSLContext.resultQuery(String).coerce(originalFields).fetch()`, in which you restore the projected field type information. I've illustrated it here: https://github.com/arsinio/jOOQ-mcve/pull/1

But luckily, there's an answer for your *actual* question, which is how to do this:

> Sometimes we modify the SQL to allow for backward compatibility with removed fields in our database (so we don't need to upload new binaries for older releases).

Better use an `ExecuteListener` for that. Using an `ExecuteListener`, you can intercept the executed query and patch it more transparently for jOOQ, i.e. jOOQ won't even notice that you changed anything in the query.

> That procedure (not included in the MCVE) shouldn't affect this issue and was removed from the MCVE for clarity.

It's a classic case of the https://xyproblem.info :) It *is* the actual issue, and I'm hoping we'll address the actual problem in the near future, versioned schema support: https://github.com/jOOQ/jOOQ/issues/4232

**Comment by arsinio** [comment]
I see, interesting...I didn't realize how closely tied the generation of the actual SQL statement was to parsing the result. In my mind it was more of:

`jOOQ statement objects` -> `generation using jOOQ context/settings` -> `SQL`
then
`result` -> `parsing using jOOQ context/settings` -> `jOOQ result objects`

where the generation and parsing were more independent (but using the same mechanisms based on the jOOQ configuration). Continuing along those lines, you should be able to parse the field from the result because the field was used to generate the original statement (and you're ostensibly dealing with the same SQL dialect)


The actual implementation makes sense though...the field itself doesn't know how it was rendered.

Looking forward to the versioned schema.


As always, thanks for your amazing work on this project!

**Comment by lukaseder** [comment]
> Continuing along those lines, you should be able to parse the field from the result because the field was used to generate the original statement (and you're ostensibly dealing with the same SQL dialect)

Define "because the field was used", by looking at your code:

```java
String sql = ctx.select(field_test)
                        .from(TEST)
                        .getSQL(ParamType.NAMED_OR_INLINED);

Result<Record> res = ctx.fetch(sql);
```

You, the developer, can "see" that these two things (generating the `String sql` value and executing it) are closely related, because they're lexically close in the same program. But how does a program *formally* prove this? By creating a huge cache that remembers every SQL string and every query ever created and matching them?

Not just that, the code that you omitted for brevity was actually *patching* said SQL string, so you're probably not even executing the same string that jOOQ generated. You say "parse", but we're not doing anything like that. "Plain SQL" means jOOQ doesn't know or care what SQL you're executing. That's both faster and allows for more vendor specific things that jOOQ can't represent.

Anyway, with the `ExecuteListener` you're intercepting the String just before jOOQ sends it to JDBC without jOOQ knowing that you modified it, so that much better suits your needs for now.

Another option that is frequently used is to implement a compatibility layer in the database itself using views, and never query the tables directly...

**Comment by arsinio** [comment]
>You, the developer, can "see" that these two things (generating the String sql value and executing it) are closely related, because they're lexically close in the same program. But how does a program formally prove this? By creating a huge cache that remembers every SQL string and every query ever created and matching them?

For a simple example like the MCVE, I don't know that the program necessarily needs to prove anything (or really cache anything). jOOQ rendered out the name of the field once using the `DSLContext` to generate the `SELECT` statement (I'm probably oversimplifying here, but I'm assuming `DSLContext` contains or can reference everything needed to create your statement in accordance with your dialect, features, etc). 

The name of that rendered field _is in_ the `Result` and the `Record` object both of which still have a reference to the `DSLContext`...so in theory you have what you need to render that field name again (in accordance with your dialect, features, etc) in a way that should match the originally rendered field for your `SELECT` statement. That would then allow you to retrieve the field without the exception.

<br><br>

>Not just that, the code that you omitted for brevity was actually patching said SQL string, so you're probably not even executing the same string that jOOQ generated.

I didn't include that code because it doesn't affect the MCVE. My point is just that code I submitted above _seems_ like it should work, and is close enough to working that I thought it might be bug. How it's being used is beside the point.

<br><br>

That all being said, you're way more in the weeds here than I am so there are probably a million design decisions that I am missing. I'm just trying to take the opportunity to further my understanding a bit!

**Comment by lukaseder** [comment]
> For a simple example like the MCVE, I don't know that the program necessarily needs to prove anything (or really cache anything). jOOQ rendered out the name of the field once using the `DSLContext` to generate the `SELECT` statement (I'm probably oversimplifying here, but I'm assuming `DSLContext` contains or can reference everything needed to create your statement in accordance with your dialect, features, etc).

There was no "name of the field". jOOQ certainly didn't generate one. That's the database's work, and jOOQ reverse engineered it from JDBC `ResultSetMetaData`. You didn't provide one in your query. You could have: `greatest(...).as("my_name")` would have worked and is usually a great idea. But the point of my screenshots is to show you that you didn't, and there isn't a well-defined, reliable name of an expression.

> (I'm probably oversimplifying here, but I'm assuming `DSLContext` contains or can reference everything needed to create your statement in accordance with your dialect, features, etc).

Yes, we could reverse engineer the weird details of auto-naming expressions in all the SQL dialects, but at what price? It's too brittle, and not even reliable as shown by the SQL Server example, where the column doesn't report a name. Worse, here's BigQuery:

![image](https://user-images.githubusercontent.com/734593/124014662-18656080-d9e4-11eb-878e-2acce35c1b87.png)

The generated field names are enumerated! There's no way we can connect the position of a field to the semantics of the field, without *knowing* the query. And that information has been lost in your program, between the render call and the fetch call.

Now look at this! H2 this time

![image](https://user-images.githubusercontent.com/734593/124016951-6f6c3500-d9e6-11eb-98f1-5ba3b32923c4.png)

What are we going to do with that?

There are chances that a "reasonable" approximation can be pulled off for some dialects, but it would break as soon as the assumptions about the auto-naming are wrong, including:

- some missing special character handling
- unexpected maximum column width
- dependency on some system setting, which we would have to mimick
- the auto-naming being undefined
- vendors changing the auto-naming behaviour in new versions

It's the old ever occurring problem in software:

- we had some well-defined, structured data (the `Query` and its projection)
- we didn't hold on to the well-defined, structured data
- we need to retrieve the well-defined, structured data again
- instead of refactoring, we're tempted to reverse engineer the well-defined, structured data from some unspecified text format, which is meant for human consumption, not machine consumption.
 
It's like storing some random localised version of YYYY-MON-DAY or DD/MM/YY or whatever the user manually enters instead of an actual `DATE` type, and then trying to pull off the guesswork with some algorithm. It never works.

> The name of that rendered field _is in_ the `Result` and the `Record` object both of which still have a reference to the `DSLContext`...so in theory you have what you need to render that field name again (in accordance with your dialect, features, etc) in a way that should match the originally rendered field for your `SELECT` statement. That would then allow you to retrieve the field without the exception.

And then, someone calls `Record.detach()`, and suddenly, the context is gone and we can no longer reliably match the database generated name with the jOOQ `Field` expression. The whole idea is just too brittle.

I hope this makes sense? Please always treat the name of an expression as undefined, unless you alias it explicitly

**Comment by arsinio** [comment]
Wow...yeah, this makes a lot of sense...thank you for taking the time to explain!
